hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/ProvidesHoodieConfig.scala

```diff
@@ -528,6 +536,40 @@ object ProvidesHoodieConfig {
       filterNullValues(overridingOpts)
   }
 
+  /**
+   * @param tableConfigKeyGeneratorClassName     key generator class name in the table config.
+   * @param partitionFieldNamesWithoutKeyGenType partition field names without key generator types
+   *                                             from the table config.
+   * @param catalogTable                         HoodieCatalogTable instance to fetch table properties.
+   * @return the write config value to set for "hoodie.datasource.write.partitionpath.field".
+   */
+  def getPartitionPathFieldWriteConfig(tableConfigKeyGeneratorClassName: String,
+                                       partitionFieldNamesWithoutKeyGenType: String,
+                                       catalogTable: HoodieCatalogTable): String = {
+    if (StringUtils.isNullOrEmpty(tableConfigKeyGeneratorClassName)) {
+      partitionFieldNamesWithoutKeyGenType
+    } else {
```

Comment:
Should we instead directly infer from the passed string `tableConfigKeyGeneratorClassName`? I mean if the string has no `:` then return `partitionFieldNamesWithoutKeyGenType`. I am not following why `tableConfigKeyGeneratorClassName` being null or empty means partition field names are without keygen type. Suppose, in a future release we drop the keygen config from table properties, then will this hold true?

Reply:
For the custom key generator, we have to take it from the properties stored in the Spark catalog table.  `partitionFieldNamesWithoutKeyGenType` is derived from the existing table configs, which can be wrong.  Also, in some code path, `tableConfigKeyGeneratorClassName` is not passed in.

Reply:
Got it.

Reply:
Flink writer should provide the correct partition field write config.  The query side may have some gaps.

Created [HUDI-7613](https://issues.apache.org/jira/browse/HUDI-7613) as a follow-up.